src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala

```diff
@@ -0,0 +1,119 @@
+import mesosphere.marathon.metrics.{Metrics, ServiceMetric}
+
+/**
+  * This filter replaces the default I/O streams with proxies that count
+  * the number of bytes that are sent to and from the client.
+  */
+class HTTPMetricsFilter extends Filter {
+  private[this] val inputBytesMetric = Metrics.counter(ServiceMetric, getClass, "inputBytes")
+  private[this] val outputBytesMetric = Metrics.counter(ServiceMetric, getClass, "outputBytes")
+
+  /**
+    * Wraps a `ServletOutputStream` and overrides the `write` method in
+    * order to count the number of bytes that went through the stream.
+    *
+    * @param upstream The stream to proxy the output to
+    */
+  private class OutputStreamCounter(val upstream: ServletOutputStream) extends ServletOutputStream {
+    var bytes: Int = 0
+
+    override def isReady: Boolean = upstream.isReady
+    override def setWriteListener(writeListener: WriteListener): Unit = upstream.setWriteListener(writeListener)
+    override def write(b: Int): Unit = {
+      bytes += 1
+      upstream.write(b)
+    }
+
+    override def write(b: Array[Byte]): Unit = {
+      bytes += b.length
+      upstream.write(b)
+    }
+
+    override def write(b: Array[Byte], off: Int, len: Int): Unit = {
+      bytes += len
+      upstream.write(b, off, len)
+    }
+  }
+
+  private class ResponseCounterWrapper(val r: HttpServletResponse) extends HttpServletResponseWrapper(r) {
+    private lazy val stream = new OutputStreamCounter(super.getOutputStream)
+    def bytes: Int = stream.bytes
+
+    /**
+      * Override the default implementation in order to return the custom stream counter
+      * @return The stream counter that wraps the base stream
+      */
+    override def getOutputStream: ServletOutputStream = stream
+  }
+
+  /**
+    * Wraps a `ServletOutputStream` and overrides the `write` method in
+    * order to count the number of bytes that went through the stream.
+    *
+    * @param upstream The stream to proxy the output to
+    */
+  private class InputStreamCounter(val upstream: ServletInputStream) extends ServletInputStream {
+    var bytes: Int = 0
+
+    override def isReady: Boolean = upstream.isReady
+    override def setReadListener(readListener: ReadListener): Unit = upstream.setReadListener(readListener)
+    override def isFinished: Boolean = upstream.isFinished
+    override def read(): Int = {
+      bytes += 1
+      upstream.read()
+    }
+
+    override def read(b: Array[Byte]): Int = {
+      bytes += b.length
+      upstream.read(b)
+    }
+
+    override def read(b: Array[Byte], off: Int, len: Int): Int = {
+      bytes += len
+      upstream.read(b, off, len)
+    }
+  }
+
+  private class RequestCounterWrapper(val r: HttpServletRequest) extends HttpServletRequestWrapper(r) {
+    private lazy val stream = new InputStreamCounter(super.getInputStream)
+    def bytes: Int = stream.bytes
```

Comment:
Different name, perhaps? `bytesRead`, `bytesWritten`.

Reply:
Not a bad idea 👍